src/ray/core_worker/task_manager.cc

```diff
@@ -1445,16 +1441,11 @@ void TaskManager::MarkTaskRetryOnResubmit(TaskEntry &task_entry) {
 
 void TaskManager::MarkTaskRetryOnFailed(TaskEntry &task_entry,
                                         const rpc::RayErrorInfo &error_info) {
+  RAY_CHECK(task_entry.IsPending());
+
   // Record the old attempt status as FAILED.
-  RAY_UNUSED(task_event_buffer_.RecordTaskStatusEventIfNeeded(
-      task_entry.spec.TaskId(),
-      task_entry.spec.JobId(),
-      task_entry.spec.AttemptNumber(),
-      task_entry.spec,
-      rpc::TaskStatus::FAILED,
-      /* include_task_info */ false,
-      worker::TaskStatusEvent::TaskStateUpdate(error_info)));
-  task_entry.MarkRetryOnFailed();
+  SetTaskStatus(task_entry, rpc::TaskStatus::FAILED, error_info);
```

Comment:
Curious to know, why we don't need to record the task status event with the change in the retry on failed case? Is it because it is already recorded as well?

Reply:
`SetTaskStatus` internally records the task status event.